### PR TITLE
fix: Also cleanup version metadata if expiring

### PR DESCRIPTION
**Pull request by juliusknorr**
Found while writing e2e tests for #39140 

Steps to reproduce:

- Upload some file multiple times with close timestamps
- Run cron which expires versions (ref https://github.com/nextcloud/server/blob/26c7cda5899d88470cf6e73fb7003c0ee0df05ac/apps/files_versions/lib/Storage.php#L87-L100)

This leads to the version files being deleted while the oc_files_versions table still contains references, meaning that they show up as broken files in the versions sidebar.

Ideally expiration should also be handled by the version manager, but this requires some larger efforts and is not backportable.

Quick script to reproduce:
```bash
#!/bin/bash
NC_URL=https://admin:admin@nextcloud.local/remote.php/webdav/

filename=$(date | md5sum | cut -d " " -f 1)
echo $filename

echo a1 > 1/$filename.md
sleep 2
echo a2 > 2/$filename.md
sleep 2
echo a3 > 3/$filename.md
sleep 2
echo a4 > 4/$filename.md
sleep 2
echo a5 > 5/$filename.md

curl -T "1/$filename.md" "$NC_URL/$filename.md"
curl -T "2/$filename.md" "$NC_URL/$filename.md"
curl -T "3/$filename.md" "$NC_URL/$filename.md"
curl -T "4/$filename.md" "$NC_URL/$filename.md"
curl -T "5/$filename.md" "$NC_URL/$filename.md"
```

Afterwards you can compare the following sql queries before and after cron for validation:

```sql
# with the echoed filename
select * from oc_filecache where name like '3882259fffd1ef36e300511488e561be%';
# with the file id
select * from oc_files_versions where file_id = 8740;
```

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

File: apps/files_versions/lib/Storage.php [hunk 1]
```diff
@@ -914,7 +914,21 @@ public static function expire($filename, $uid) {
...
 				\OC_Hook::emit('\OCP\Versions', 'preDelete', ['path' => $path, 'trigger' => self::DELETE_TRIGGER_QUOTA_EXCEEDED]);
+
+				// Make sure to cleanup version table relations as expire does not pass deleteVersion
+				try {
+					/** @var VersionsMapper $versionsMapper */
+					$versionsMapper = \OC::$server->get(VersionsMapper::class);
+					$file = \OC::$server->get(IRootFolder::class)->getUserFolder($uid)->get($filename);
+					$pathparts = pathinfo($path);
+					$timestamp = (int)substr($pathparts['extension'] ?? '', 1);
+					$version = $versionsMapper->findVersionForFileId($file->getId(), $timestamp);
```
Comment (juliushaertl) [inline apps/files_versions/lib/Storage.php:925]: We should probably also skip the expire if the version has a label

Comment (artonge) [reply apps/files_versions/lib/Storage.php:925]: Yes, that would be expected, as in:

https://github.com/nextcloud/server/blob/c3c58b66f2f60a9191d673cee3c2544a23acdfdf/apps/files_versions/lib/Storage.php#L595-L607

Comment (juliushaertl) [reply apps/files_versions/lib/Storage.php]: Pushed.

**Review by artonge — APPROVED** [review]

**Review by icewind1991 — APPROVED** [review]

**Comment by solracsf** [comment]
@juliushaertl no 25 ?

**Comment by juliushaertl** [comment]
Only 26+ is affected where the new table for metadata was introduced

**Pull request merged**